File: frontend/lib/route-access.ts
import type { User } from "@/lib/api";
import { canSeeNavItem, isOrgAdmin, type NavAudience, viewerReadOnlyMessage } from "@/lib/nav-access";
import { canAccessPlatformPath } from "@/lib/platform-access";
import {
  isOrgFeatureEnabled,
  type OrgFeatureFlagKey,
} from "@/lib/use-org-feature-flags";

type RouteRule = {
  prefix: string;
  audience: NavAudience | NavAudience[];
  excludeViewers?: boolean;
  featureFlag?: OrgFeatureFlagKey;
};

const ROUTE_RULES: RouteRule[] = [
  { prefix: "/trees/new", audience: "can_write" },
  { prefix: "/field-ops", audience: ["professional", "field_supervisor"], excludeViewers: true },
  { prefix: "/projects", audience: ["professional", "field_supervisor", "field_worker"] },
  {
    prefix: "/portfolio-health",
    audience: ["professional", "field_supervisor"],
    featureFlag: "satellite",
  },
  {
    prefix: "/monitoring",
    audience: ["professional", "field_supervisor"],
    featureFlag: "satellite",
  },
  {
    prefix: "/intelligence",
    audience: ["professional", "field_supervisor"],
    featureFlag: "satellite",
  },
  {
    prefix: "/satellite",
    audience: ["professional", "field_supervisor"],
    featureFlag: "satellite",
  },
  { prefix: "/bioacoustic", audience: "professional", featureFlag: "bioacoustic" },
  { prefix: "/reports", audience: ["professional", "field_supervisor"], featureFlag: "reports" },
  { prefix: "/assistant", audience: "all", featureFlag: "ai_scan" },
  { prefix: "/settings/team", audience: "org_admin" },
];

function matchingRule(pathname: string): RouteRule | undefined {
  return ROUTE_RULES.find(
    (rule) => pathname === rule.prefix || pathname.startsWith(`${rule.prefix}/`),
  );
}

export function canAccessPath(
  user: User | null | undefined,
  pathname: string,
  featureFlags?: Map<OrgFeatureFlagKey, boolean>,
): boolean {
  if (!user) return false;

  if (pathname.startsWith("/platform")) {
    return canAccessPlatformPath(user, pathname);
  }

  const rule = matchingRule(pathname);
  if (rule) {
    if (!canSeeNavItem(user, rule.audience, { excludeViewers: rule.excludeViewers })) {
      return false;
    }
    if (rule.featureFlag && !isOrgFeatureEnabled(featureFlags, rule.featureFlag)) {
      return false;
    }
    return true;
  }

  return true;
}

export type RouteAccessMessageKey =
  | "teamAdmin"
  | "fieldOps"
  | "platform"
  | "default"
  | "treesNew"
  | "featureDisabled";

export function routeAccessDeniedKey(pathname: string): RouteAccessMessageKey {
  if (pathname.startsWith("/settings/team")) return "teamAdmin";
  if (pathname.startsWith("/trees/new")) return "treesNew";
  if (pathname.startsWith("/field-ops")) return "fieldOps";
  if (pathname.startsWith("/platform")) return "platform";
  return "default";
}

export function routeAccessDeniedFeatureKey(
  pathname: string,
  featureFlags?: Map<OrgFeatureFlagKey, boolean>,
): OrgFeatureFlagKey | undefined {
  const rule = matchingRule(pathname);
  if (!rule?.featureFlag) return undefined;
  if (isOrgFeatureEnabled(featureFlags, rule.featureFlag)) return undefined;
  return rule.featureFlag;
}

/** @deprecated Use routeAccessDeniedKey with useTranslations("access") in UI */
export function routeAccessDeniedMessage(pathname: string): string {
  if (pathname.startsWith("/settings/team")) {
    return "Organization admin access is required to manage your team.";
  }
  if (pathname.startsWith("/trees/new")) {
    return viewerReadOnlyMessage("trees");
  }
  if (pathname.startsWith("/field-ops")) {
    return "Field operations are limited to supervisors and program leads.";
  }
  if (pathname.startsWith("/platform")) {
    return "Platform administration access is required.";
  }
  return "Your role does not include access to this section.";
}

File: mobile/lib/src/route_access.dart
import 'nav_access.dart' show UserMap, canSeeNavItem;

class _RouteRule {
  const _RouteRule(this.prefix, this.audience, {this.excludeViewers = false});

  final String prefix;
  final Object audience;
  final bool excludeViewers;
}

const _routeRules = [
  _RouteRule('/trees/new', 'can_write'),
  _RouteRule('/projects', ['professional', 'field_supervisor', 'field_worker']),
  _RouteRule('/bioacoustic', 'professional'),
  _RouteRule('/field-ops', ['professional', 'field_supervisor', 'field_worker'], excludeViewers: true),
  _RouteRule('/monitoring', ['professional', 'field_supervisor']),
  _RouteRule('/map/draw', 'can_write'),
  _RouteRule('/reports', ['professional', 'field_supervisor']),
  _RouteRule('/credits', ['professional', 'field_supervisor']),
  _RouteRule('/carbon', 'all'),
];

bool canAccessPath(UserMap? user, String path) {
  if (user == null) return false;

  // Survival / re-geotag requires write access.
  if (RegExp(r'^/trees/[^/]+/survival').hasMatch(path)) {
    return canSeeNavItem(user, 'can_write');
  }

  for (final rule in _routeRules) {
    if (path == rule.prefix || path.startsWith('${rule.prefix}/')) {
      return canSeeNavItem(
        user,
        rule.audience,
        excludeViewers: rule.excludeViewers,
      );
    }
  }
  return true;
}

String routeAccessDeniedMessage(String path) {
  if (path.startsWith('/trees/new') || RegExp(r'^/trees/[^/]+/survival').hasMatch(path)) {
    return 'Your viewer role is read-only. Ask your program manager for write access.';
  }
  if (path.startsWith('/projects')) {
    return 'Projects are limited to field teams and program members.';
  }
  if (path.startsWith('/bioacoustic')) {
    return 'Bioacoustic monitoring requires a professional program.';
  }
  if (path.startsWith('/field-ops')) {
    return 'Field operations are limited to field teams and program leads.';
  }
  if (path.startsWith('/monitoring')) {
    return 'Monitoring is limited to supervisors and program members.';
  }
  if (path.startsWith('/reports')) {
    return 'Reports are limited to supervisors and program members.';
  }
  if (path.startsWith('/credits')) {
    return 'Credit ledgers are limited to program members.';
  }
  if (path.startsWith('/map/draw')) {
    return 'Drawing work areas requires write access.';
  }
  return 'Your role does not include access to this section.';
}
